rust/src/driver_manager.rs

```diff
@@ -0,0 +1,892 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Load and use ADBC drivers.
+//!
+//! ## Loading a driver
+//!
+//! Drivers are initialized using a function provided by the driver as a main
+//! entrypoint, canonically called `AdbcDriverInit`. (Although many will use a
+//! different name to support statically linking multiple drivers within the
+//! same program.)
+//!
+//! To load from a function, use [AdbcDriver::load_from_init].
+//!
+//! To load from a dynamic library, use [AdbcDriver::load].
+//!
+//! ## Using across threads
+//!
+//! [AdbcDriver] and [AdbcDatabase] can be used across multiple threads. They
+//! hold their inner implementations within [std::sync::Arc], so they are
+//! cheaply copy-able.
+//!
+//! [AdbcConnection] should not be used across multiple threads. Driver
+//! implementations do not guarantee connection APIs are safe to call from
+//! multiple threads, unless calls are carefully sequenced. So instead of using
+//! the same connection across multiple threads, create a connection for each
+//! thread. [AdbcConnectionBuilder] is [core::marker::Send], so it can be moved
+//! to a new thread before initialized into an [AdbcConnection]. [AdbcConnection]
+//! holds it's inner data in a [std::rc::Rc], so it is also cheaply copyable.
```

Comment:
I was thinking just last night that we might want to implement async `execute` methods, which would involve invoking with `spawn_blocking()`. That might mean we need to make `AdbcConnection` `Send`, but not 100% sure yet. Maybe something for a follow up.

Reply:
I've wanted to implement async interfaces in general. That'll require some thought between Flight RPC, and Rust/C++/Go. (I want to make the C-level ABI in a way that all three languages can expose their native concurrency models efficiently, and import others' models efficiently. But that means I need to really go study up on how those models work.) I'd appreciate any thoughts you and @zeroshade have. 

e.g.: gRPC C++ uses a callback-based model. UCX (in C) uses a polling-based model. There's precedent like kqueue/epoll, and of course the new io_uring (don't know if that's relevant). 

There's also Python's model to consider (as a consumer), and probably we'd want to anticipate things like OpenTelemetry (or generally: non-stack-frame, non-thread based 'task' context). And eventually, whenever I get around to JNI bindings, Java's model as well (which, with virtual threading, is also getting more complicated/interesting).

Reply:
Oh, and consideration of whether we need an async version of C Data Interface...

Reply:
FWIW the Rust model is described well here: https://rust-lang.github.io/async-book/02_execution/02_future.html. 

But for now I was simply thinking that Rust driver manager could wrap the blocking calls into C drivers with [tokio::spawn_blocking()](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html), which sends the call to another thread to execute. This is what Rust async libraries do for filesystem interaction on systems that don't provide a native async interface. And then native Rust drivers could always be async. But maybe there is value in having both async and blocking versions of `execute()` in the Rust API. Something I'll try to get feedback from the Rust community on once I have a proposal for the ADBC Rust API.

Reply:
Cool. If I can find some spare time...

I think async interfaces in general throughout our ecosystem are going to be more important as they get adopted more by developers (C++ is, well, coming along, but Go/mostly Rust/sorta Python have generally mature concurrency stories nowadays)